pkg/tfvars/baremetal/baremetal.go

```diff
@@ -0,0 +1,107 @@
+// Package baremetal contains bare metal specific Terraform-variable logic.
+package baremetal
+
+import (
+	"encoding/json"
+	"github.com/metal3-io/baremetal-operator/pkg/bmc"
+	"github.com/metal3-io/baremetal-operator/pkg/hardware"
+	libvirttfvars "github.com/openshift/installer/pkg/tfvars/libvirt"
+	"github.com/openshift/installer/pkg/types/baremetal"
+	"github.com/pkg/errors"
+)
+
+type config struct {
+	LibvirtURI         string `json:"libvirt_uri,omitempty"`
+	IronicURI          string `json:"ironic_uri,omitempty"`
+	Image              string `json:"os_image,omitempty"`
+	ExternalBridge     string `json:"external_bridge,omitempty"`
+	ProvisioningBridge string `json:"provisioning_bridge,omitempty"`
+
+	// Data required for control plane deployment - several maps per host, because of terraform's limitations
+	Hosts         []map[string]interface{} `json:"hosts"`
+	RootDevices   []map[string]interface{} `json:"root_devices"`
+	Properties    []map[string]interface{} `json:"properties"`
+	DriverInfos   []map[string]interface{} `json:"driver_infos"`
+	InstanceInfos []map[string]interface{} `json:"instance_infos"`
+}
+
+// TFVars generates bare metal specific Terraform variables.
+func TFVars(libvirtURI, ironicURI, osImage, externalBridge, provisioningBridge string, platformHosts []*baremetal.Host, image baremetal.Image) ([]byte, error) {
+	osImage, err := libvirttfvars.CachedImage(osImage)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to use cached libvirt image")
+	}
+
+	var hosts, rootDevices, properties, driverInfos, instanceInfos []map[string]interface{}
+
+	for _, host := range platformHosts {
+		// Get hardware profile
+		if host.HardwareProfile == "default" {
+			host.HardwareProfile = hardware.DefaultProfileName
+		}
+
+		profile, err := hardware.GetProfile(host.HardwareProfile)
+		if err != nil {
+			return nil, err
+		}
+
+		// BMC Driver Info
+		accessDetails, _ := bmc.NewAccessDetails(host.BMC.Address)
+		credentials := bmc.Credentials{
+			Username: host.BMC.Username,
+			Password: host.BMC.Password,
+		}
+		driverInfo := accessDetails.DriverInfo(credentials)
+		driverInfo["deploy_kernel"] = image.DeployKernel
+		driverInfo["deploy_ramdisk"] = image.DeployRamdisk
+
+		// Host Details
+		hostMap := map[string]interface{}{
+			"name":         host.Name,
+			"port_address": host.BootMACAddress,
+			"driver":       accessDetails.Type(),
+		}
+
+		// Properties
+		propertiesMap := map[string]interface{}{
+			"local_gb": profile.LocalGB,
+			"cpu_arch": profile.CPUArch,
+		}
+
+		// Root device hints
+		rootDevice := make(map[string]interface{})
+		if profile.RootDeviceHints.HCTL != "" {
+			rootDevice["hctl"] = profile.RootDeviceHints.HCTL
+		} else {
+			rootDevice["name"] = profile.RootDeviceHints.DeviceName
```

Comment:
This feels sort of brittle. What if we ever use the other options? vendor,
wwn_vendor_extension, wwn_with_extension, by_path, serial, wwn,
size, rotational, name, hctl, model
Now I have to use it in the metal3 codebase and come out here and rewrite the tfvar translator. Do we think this would make sense as a helper function in metal3?

Reply:
in the install_ipi.md file, the whole hardwareProfile thing is highlighted as an area particularly subject to change while this is still experimental

indeed, some helper code might improve this some, but i'm not even sure we got the right abstraction defined yet, but it seems ok while this is experimental